govcd/nsxv_distributed_firewall.go

```diff
@@ -0,0 +1,486 @@
+	// header "If-Match" with the contents of the ETag from a previous read.
+	// The same data can be found in the "GenerationNumber" within the section to update.
+	// The value of the ETag changes at every GET
+	if dfw.Etag == "" && firewallSection.GenerationNumber != "" {
+		dfw.Etag = firewallSection.GenerationNumber
+	}
+	config.Layer3Sections = &types.Layer3Sections{Section: &firewallSection}
+	dfw.Configuration = &config
+	dfw.Configuration.Layer3Sections = config.Layer3Sections
+	dfw.enabled = true
+	return &config, nil
+}
+
+// IsEnabled returns true when the distributed firewall is enabled
+func (dfw *NsxvDistributedFirewall) IsEnabled() (bool, error) {
+	if dfw.VdcId == "" {
+		return false, fmt.Errorf("no VDC set for this NsxvDistributedFirewall")
+	}
+
+	conf, err := dfw.GetConfiguration()
+	if err != nil {
+		return false, nil
+	}
+	if dfw.client.APIVersion == "36.0" {
+		return conf != nil, nil
+	}
+	return true, nil
+}
+
+// Enable makes the distributed firewall available
+// It fails with a non-NSX-V VDC
+func (dfw *NsxvDistributedFirewall) Enable() error {
+	dfw.enabled = false
+	if dfw.VdcId == "" {
+		return fmt.Errorf("no AdminVdc set for this NsxvDistributedFirewall")
+	}
+	initialUrl, err := dfw.client.buildUrl("network", "firewall", "vdc", extractUuid(dfw.VdcId))
+	if err != nil {
+		return err
+	}
+
+	requestUrl, err := url.ParseRequestURI(initialUrl)
+	if err != nil {
+		return err
+	}
+
+	req := dfw.client.NewRequest(nil, http.MethodPost, *requestUrl, nil)
+
+	resp, err := checkResp(dfw.client.Http.Do(req))
+	if err != nil {
+		return err
+	}
+	if resp != nil && resp.StatusCode != http.StatusCreated {
+		return fmt.Errorf("[enable DistributedFirewall] expected status code %d - received %d", http.StatusCreated, resp.StatusCode)
+	}
+	dfw.enabled = true
+	return nil
+}
+
+// Disable removes the availability of a distributed firewall
+// WARNING: it also removes all rules
+func (dfw *NsxvDistributedFirewall) Disable() error {
+	if dfw.VdcId == "" {
+		return fmt.Errorf("no AdminVdc set for this NsxvDistributedFirewall")
+	}
+	initialUrl, err := dfw.client.buildUrl("network", "firewall", "vdc", extractUuid(dfw.VdcId))
+	if err != nil {
+		return err
+	}
+
+	requestUrl, err := url.ParseRequestURI(initialUrl)
+	if err != nil {
+		return err
+	}
+
+	req := dfw.client.NewRequest(nil, http.MethodDelete, *requestUrl, nil)
+
+	resp, err := checkResp(dfw.client.Http.Do(req))
+	if err != nil {
+		// VCD 10.3.x sometimes returns an error even though the removal succeeds
+		if dfw.client.APIVersion == "36.0" {
+			conf, _ := dfw.GetConfiguration()
+			if conf == nil {
+				return nil
+			}
+		}
+		return fmt.Errorf("error deleting Distributed firewall: %s", err)
+
+	}
+	if resp != nil && resp.StatusCode != http.StatusNoContent {
+		return fmt.Errorf("[disable DistributedFirewall] expected status code %d - received %d", http.StatusNoContent, resp.StatusCode)
+	}
+	dfw.Configuration = nil
+	dfw.Services = nil
+	dfw.ServiceGroups = nil
```

Comment:
Does it mean we remove all services and service groups as well?

Reply:
what we remove is the cache associated to the DFW object